### PR TITLE
[WIP] Experiement Golang SDK

### PR DESCRIPTION
**Pull request by bianpengyuan**
This is some exploratory work I recently did. Lots of works remain, such as comment and test. Not sure if this worths putting more time into considering the WASI shifting. This is based on tinygo, which not yet fully supports reflect thus cannot do protobuf, but I think they are on the way there.

Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

### CODE REVIEW

File: source/extensions/common/wasm/wasm.cc [hunk 1]
```diff
@@ -2211,6 +2260,16 @@ void Wasm::registerCallbacks() {
 
   _REGISTER_PROXY(setEffectiveContext);
 #undef _REGISTER_PROXY
+
+#define _REGISTER_GOLANG_PROXY(_fn)                                                                  \
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm.cc]: I'm confused by why you need those? C++, Rust and TinyGo should be all calling the same functions.

Comment (bianpengyuan) [reply source/extensions/common/wasm/wasm.cc]: tinygo imports function with package name as prefix (`envoywasmsdk`) and two extra params appended at the end: https://tinygo.org/compiler-internals/calling-convention/.

Comment (bianpengyuan) [reply source/extensions/common/wasm/wasm.cc]: This is wat of tinygo generated wasm file:

``` 
(import "env" "envoywasmsdk._proxy_getHeaderMapValue" (func $envoywasmsdk._proxy_getHeaderMapValue (type $t2)))
  (import "env" "envoywasmsdk._proxy_log" (func $envoywasmsdk._proxy_log (type $t3)))
  (import "env" "envoywasmsdk._proxy_addHeaderMapValue" (func $envoywasmsdk._proxy_addHeaderMapValue (type $t2)))
```

Comment (ly-2016) [reply source/extensions/common/wasm/wasm.cc]: https://github.com/tinygo-org/tinygo/pull/455

**Comment by Shikugawa** [comment]
How's it going this PR?

**Comment by zhongfox** [comment]
How's it going this PR?

**Comment by dinusha92** [comment]
Is there any update on this? :)

**Comment by PiotrSikora** [comment]
@Shikugawa @zhongfox @dinusha92 I don't believe anybody from our team is actively working on the Proxy-Wasm Go SDK (either TinyGo or regular Go), but we could provide some guidance and/or code reviews if you're interested in working on it. 

cc @bianpengyuan @incfly @jplevyak

File: source/extensions/common/wasm/wasm.cc (hunk 1 above)
Comment (aykevl) [reply source/extensions/common/wasm/wasm.cc]: I just came across this PR. Please note that the calling convention should be considered an implementation detail. If you want a stable calling convention, use `//export` which is intended to match the C calling convention.

**Comment by mathetake** [comment]
hi folks, I personally developed go sdk for proxy-wasm : https://github.com/mathetake/proxy-wasm-go

This is still bare minimum but it works fine within the realm of basic examples without modifying envoryproxy's host implementations.

**Comment by hobbytp** [comment]
Any update? thanks,

**Comment by PiotrSikora** [comment]
> Any update? thanks,

You should use https://github.com/mathetake/proxy-wasm-go.

**Pull request closed without merging**